stack/maxima/contrib/linearalgebra.mac

```diff
@@ -0,0 +1,1205 @@
+    cond2: ev(lmax(svs)/lmin(svs),simp)
+  ),
+  return(cond2)
+);
+
+/** 
+ * Solve the matrix equation Ax = b given matrix A and column vector (or list) b.
+ * Optionally will find a least squares solution
+ * Always returns a general solution if one exists, even in the least squares case
+ * If a single solution is required, use pseudoinverse(A) . b instead.
+ * 
+ * @param[matrix] A An mxn matrix
+ * @param[matrix] b A mx1 matrix (or a list with m entries)
+ * @param[boolean] lstsq Optional: if given true then a least squares solution will be obtained. If false or omitted, only exact solutions obtained.
+ * @return[matrix] The general solution to Ax = b. If no solution exists and lstsq is not true, then matrix([]) is returned.
+ */
+mat_solve(A,b,[lstsq]):= block([m,n,vars,eqns,sol],
+  if emptyp(lstsq) then lstsq: false else lstsq:first(lstsq),
+  if listp(b) then b: transpose(b),
+  [m, n]: matrix_size(A),
+  if ev(is(first(matrix_size(b))#m),simp) then return(matrix([])),
+  vars: rest(stack_var_makelist(tmp,n)),
+  if lstsq then AT: transpose(A) else AT: ident(m),
+  eqns: list_matrix_entries(ev((AT . A) . transpose(vars) - (AT . b),simp)),
+  sol: map(rhs,linsolve(eqns,vars)),
+  if emptyp(sol) then return(matrix(sol)) else return(transpose(matrix(sol)))
+);
+
+/** 
+ * Given a list of lists or a matrix, make a basis for R^m where m = length of each vector. 
+ * If you don't want to expand to R^m, use remove_dep instead (it is called here too)
+ * Optionally will make this basis orthonormal, mostly useful when the existing basis is orthogonal but not R^m
+ * Arguably "basisify" is a poor name - suggestions welcome before published
+ *
+ * @param[matrix or list] M A matrix or list of lists
+ * @param[boolean] orth Optional: if true then returned basis will be orthonormal. Note that this may remove some basis vectors if they are not already orthogonal.
+ * @return[matrix or list] A basis for R^m where m is the length of each vector in M
+ */
+
+basisify(M,[orth]):= block([ex_op,m,n,vecs,new_vecs,ii],
+  if emptyp(orth) then orth: false else orth: first(orth),
+  ex_op: "matrix",
+  if listp(M) then block(M: column_stack(M), ex_op: "list"),
+  if not(lin_indp(M)) then M: remove_dep(M),
+  [m, n]: matrix_size(M),
+  vecs: args(transpose(M)),
+  new_vecs: args(ident(m)),
+  for ii: 1 thru m do block(
+    ii: ev(ii,simp),
+    if lin_indp(append(vecs,[new_vecs[ii]])) then vecs: append(vecs,[new_vecs[ii]])
+  ),
+  if orth then block(
+    vecs: ev(gramschmidt(apply(matrix,vecs)),simp),
+    vecs: ev(map(lambda([ex],ex/sqrt(ex.ex)),vecs),simp)
+  ),
+  if is(ex_op="matrix") then return(transpose(apply(matrix,vecs))) else return(vecs)
+);
+
+/**
+ * Maps the gcd (greatest common divisor) function across a list iteratively
+ * 
+ * @param[list] ex A list of numbers
+ * @return[number] The greatest common divisor of all elements in ex
+ */
+lgcd(ex):= block([ex_gcd,ii],
```

Comment:
You could probably simplify this by using `lreduce` (or similar) instead of a for loop: https://maxima.sourceforge.io/docs/manual/maxima_21.html#lreduce